cpp/src/stream_compaction/drop_duplicates.cu

```diff
@@ -130,18 +174,93 @@ std::unique_ptr<table> drop_duplicates(table_view const& input,
                         mr);
 }
 
+std::unique_ptr<table> unordered_drop_duplicates(table_view const& input,
+                                                 std::vector<size_type> const& keys,
+                                                 null_equality nulls_equal,
+                                                 rmm::cuda_stream_view stream,
+                                                 rmm::mr::device_memory_resource* mr)
+{
+  if (0 == input.num_rows() || 0 == input.num_columns() || 0 == keys.size()) {
+    return empty_like(input);
+  }
+
+  auto keys_view = input.select(keys);
+  auto table_ptr = cudf::table_device_view::create(keys_view, stream);
+  auto has_null  = nullate::DYNAMIC{cudf::has_nulls(keys_view)};
+  auto const num_rows{table_ptr->num_rows()};
+
+  hash_map_type key_map{compute_hash_table_size(num_rows),
+                        COMPACTION_EMPTY_KEY_SENTINEL,
+                        COMPACTION_EMPTY_VALUE_SENTINEL,
+                        detail::hash_table_allocator_type{default_allocator<char>{}, stream},
+                        stream.value()};
+
+  compaction_hash hash_key{has_null, *table_ptr};
+  row_equality_comparator row_equal(has_null, *table_ptr, *table_ptr, nulls_equal);
+
+  auto iter = cudf::detail::make_counting_transform_iterator(
+    0, [] __device__(size_type i) { return cuco::make_pair(std::move(i), std::move(i)); });
+  // insert unique indices into the map.
+  key_map.insert(iter, iter + num_rows, hash_key, row_equal, stream.value());
+
+  auto counting_iter = thrust::make_counting_iterator<size_type>(0);
+  rmm::device_uvector<bool> existences(num_rows, stream, mr);
+  // enumerate all indices to check if they are present in the map.
+  key_map.contains(counting_iter, counting_iter + num_rows, existences.begin(), hash_key);
+
+  auto const output_size{key_map.get_size()};
+
+  // write unique indices to a numeric column
+  auto unique_indices = make_numeric_column(
+    data_type{type_id::INT32}, output_size, mask_state::UNALLOCATED, stream, mr);
+  auto mutable_view = mutable_column_device_view::create(*unique_indices, stream);
+  thrust::copy_if(rmm::exec_policy(stream),
+                  counting_iter,
+                  counting_iter + num_rows,
+                  existences.begin(),
+                  mutable_view->begin<size_type>(),
+                  [] __device__(bool const b) { return b; });
```

Comment:
Can this be `thrust::identity`? https://thrust.github.io/doc/structthrust_1_1identity.html